igvc_rviz_plugins/src/sensor_panel.cpp

```diff
@@ -0,0 +1,171 @@
+  sensor_timer->setInterval(INTERVAL);  // goes off every 250 milliseconds
+  // calls timer whenever sensor_timer goes off
+  QObject::connect(sensor_timer, SIGNAL(timeout()), this, SLOT(timer()));
+
+  /* Initialize our subscriber to listen to the /speed topic.
+  * Note the use of boost::bind, which allows us to give the callback a pointer to our UI label.
+  */
+  imu_sub = nh.subscribe<sensor_msgs::Imu>("/imu", 1, boost::bind(&sensor_panel::imu_callback, this, _1, labels[0]));
+  lidar_sub = nh.subscribe<sensor_msgs::PointCloud2>("/scan/pointcloud", 1,
+                                                     boost::bind(&sensor_panel::lidar_callback, this, _1, labels[1]));
+  camCenter_sub = nh.subscribe<sensor_msgs::Image>("/usb_cam_center/image_raw", 1,
+                                                   boost::bind(&sensor_panel::camCenter_callback, this, _1, labels[2]));
+  camLeft_sub = nh.subscribe<sensor_msgs::Image>("/usb_cam_left/image_raw", 1,
+                                                 boost::bind(&sensor_panel::camLeft_callback, this, _1, labels[3]));
+  camRight_sub = nh.subscribe<sensor_msgs::Image>("/usb_cam_right/image_raw", 1,
+                                                  boost::bind(&sensor_panel::camRight_callback, this, _1, labels[4]));
+  gps_sub =
+      nh.subscribe<sensor_msgs::NavSatFix>("/fix", 1, boost::bind(&sensor_panel::gps2_callback, this, _1, labels[5]));
+
+  /* Use QT layouts to add widgets to the panel.
+  * Here, we're using a VBox layout, which allows us to stack all of our widgets vertically.
+  */
+  QVBoxLayout *layout = new QVBoxLayout;
+
+  // adds all the widgets to the layout
+  for (int i = 0; i < NUMSENSORS; i++)
+  {
+    layout->addWidget(labels[i]);
+  }
+  setLayout(layout);
+  timer();  // inits to disabled
+}
+
+void sensor_panel::imu_callback(const sensor_msgs::ImuConstPtr &msg, QLabel *label)
+{
+  label->setText("IMU: Enabled" /*.c_str()*/);
+  isActive[0] = true;
+}
+
+void sensor_panel::lidar_callback(const sensor_msgs::PointCloud2ConstPtr &msg, QLabel *label)
+{
+  label->setText("Lidar: Enabled");
+  isActive[1] = true;
+}
+
+void sensor_panel::camCenter_callback(const sensor_msgs::ImageConstPtr &msg, QLabel *label)
+{
+  label->setText("Center Camera: Enabled");
+  isActive[2] = true;
+}
+
+void sensor_panel::camLeft_callback(const sensor_msgs::ImageConstPtr &msg, QLabel *label)
+{
+  label->setText("Left Camera: Enabled");
+  isActive[3] = true;
+}
+
+void sensor_panel::camRight_callback(const sensor_msgs::ImageConstPtr &msg, QLabel *label)
+{
+  label->setText("Right Camera: Enabled");
+  isActive[4] = true;
+}
+void sensor_panel::gps2_callback(const sensor_msgs::NavSatFixConstPtr &msg, QLabel *label)
+{
+  label->setText("GPS: Enabled");
+  isActive[5] = true;
+}
+
+/*
+*Sets all of the activity statuses to disabled, and sets labels to diabled if already disabled
+*/
+void sensor_panel::reset_labels()
+{
+  for (int i = 0; i < NUMSENSORS; i++)
```

Comment:
for each loop?

Reply:
Done. Pushed

Reply:
If I use a fro each loop and i have to access the labels stored in the map in the set_label() method. How can I figure out which QLabel each bool in isActive corresponds to?

Reply:
I think that these two data structures would be easiest to deal with if they had the same indexing structure. So I think unless we store the bools in a map with corresponding keys to the labels then an array would be better to use.

Reply:
I concur; I fail to see how a for-each loop would be any more convenient, considering that additional code elements are necessary to iterate over a map

Reply:
This comment was centered around getting rid of the sensor number being hardcoded.